hugr-llvm/src/sum.rs

```diff
@@ -60,114 +124,365 @@ impl<'c> LLVMSumType<'c> {
-                .into_struct_value();
-        }
-        let mut sum_v = self.get_poison().as_basic_value_enum().into_struct_value();
-        if self.has_tag_field() {
-            sum_v = builder
-                .build_insert_value(
-                    sum_v,
-                    self.get_tag_type().const_int(tag as u64, false),
-                    0u32,
-                    "",
-                )?
-                .into_struct_value();
-        }
-        Ok(builder
-            .build_insert_value(sum_v, row_v, variant_field_index as u32, "")?
-            .as_basic_value_enum())
+                value.as_basic_value_enum()
+            }
+            Self::MultiVariant {
+                field_indices,
+                variant_types,
+                value_type,
+            } => {
+                let variant_field_types = &variant_types[tag];
+                let variant_field_indices = &field_indices[tag];
+                let mut value = builder
+                    .build_insert_value(
+                        value_type.get_poison(),
+                        self.tag_type().const_int(tag as u64, false),
+                        0,
+                        "",
+                    )?
+                    .into_struct_value();
+                for (t, mb_i, v) in izip!(variant_field_types, variant_field_indices, vs) {
+                    ensure!(&v.get_type() == t);
+                    if let Some(i) = mb_i {
+                        value = builder
+                            .build_insert_value(value, v, *i as u32 + 1, "")?
+                            .into_struct_value();
+                    }
+                }
+                value.as_basic_value_enum()
+            }
+        };
+        debug_assert_eq!(value.get_type(), self.value_type());
+        Ok(value)
     }
 
     /// Get the type of the value that would be returned by `build_get_tag`.
-    pub fn get_tag_type(&self) -> IntType<'c> {
-        self.0.get_context().i32_type()
-    }
-
-    fn has_tag_field(&self) -> bool {
-        sum_type_has_tag_field(&self.1)
+    pub fn tag_type(&self) -> IntType<'c> {
+        match self {
+            Self::Void { tag_type, .. } => *tag_type,
+            Self::Unit { tag_type, .. } => *tag_type,
+            Self::NoFields { value_type, .. } => *value_type,
+            Self::SingleVariantSingleField { tag_type, .. } => *tag_type,
+            Self::SingleVariantMultiField { tag_type, .. } => *tag_type,
+            Self::MultiVariant { value_type, .. } => value_type
+                .get_field_type_at_index(0)
+                .unwrap()
+                .into_int_type(),
+        }
     }
 
-    fn get_variant_field_index(&self, tag: usize) -> usize {
-        tag + (if self.has_tag_field() { 1 } else { 0 })
+    /// The underlying LLVM type.
+    pub fn value_type(&self) -> BasicTypeEnum<'c> {
+        match self {
+            Self::Void { tag_type, .. } => (*tag_type).into(),
```

Comment:
Shouldn't this be an empty struct according to the doc on `LLVMSumTypeEnum::Void`?